packages/product-json-to-csv/src/main.js

```diff
@@ -0,0 +1,292 @@
+      state: State,
+    } => ({ state: body }))
+  }
+
+  _resolveCategories(
+    categoriesReference: Array<TypeReference>
+  ): Array<Category> | {} {
+    if (!categoriesReference || !categoriesReference.length) return {}
+
+    const categoryIds = categoriesReference.map(
+      (category: TypeReference): string => category.id
+    )
+    return this._getCategories(categoryIds).then(
+      (categories: Array<Category>): { categories: Array<Category> } => {
+        if (this.parserConfig.categoryBy !== 'namedPath') return { categories }
+        return Promise.map(categories, (cat: Category): Array<Category> =>
+          this._resolveAncestors(cat)
+        ).then((categoriesWithParents: Array<Category>): Object => ({
+          categories: categoriesWithParents,
+        }))
+      }
+    )
+  }
+
+  _resolveCategoryOrderHints(categoryOrderHintsReference: Object): Object {
+    if (
+      !categoryOrderHintsReference ||
+      !Object.keys(categoryOrderHintsReference).length
+    )
+      return { categoryOrderHints: undefined }
+
+    const catIdentifier = this.parserConfig.categoryOrderHintBy
+    const lang = this.parserConfig.language
+    const categoryIds = Object.keys(categoryOrderHintsReference)
+    return this._getCategories(categoryIds).then(
+      (resolvedCategories: Array<Category>): Object => {
+        const categoryOrderHints = {}
+        resolvedCategories.forEach((category: Category) => {
+          const catRef =
+            catIdentifier === 'name'
+              ? category[catIdentifier][lang]
+              : category[catIdentifier]
+          categoryOrderHints[catRef] = categoryOrderHintsReference[category.id]
+        })
+        return { categoryOrderHints }
+      }
+    )
+  }
+
+  _resolveAncestors(category: Category): Promise<Category> {
+    const getParent = async (cat: Category): Promise<Category> => {
+      if (!cat.parent) return cat
+
+      const resolvedCategory = await this._getCategories([cat.parent.id])
+      return { ...cat, parent: await getParent(resolvedCategory[0]) }
+    }
+    return getParent(category)
+  }
+
+  // This method decides if to get the categories from cache or API
+  _getCategories(ids: Array<string>): Promise<Array<Object>> {
+    const notCachedIds = []
+    const cachedCategories = []
+    ids.forEach((id: string) => {
+      if (this.categoriesCache[id])
+        cachedCategories.push(this.categoriesCache[id])
+      else notCachedIds.push(id)
+    })
+    if (!notCachedIds.length) return Promise.resolve(cachedCategories)
+
+    const predicate = `id in ("${notCachedIds.join('", "')}")`
+    const categoriesService = this._createService('categories')
+    const uri = categoriesService.where(predicate).build()
+    return this.fetchReferences(uri).then(
+      ({ body: { results } }: SuccessResult): Array<Category> => {
+        results.forEach((result: Category) => {
+          cachedCategories.push(result)
```

Comment:
I don't get this... Could you expatiate?

Reply:
A key-value cache is defined [here](https://github.com/commercetools/nodejs/pull/326/files/3ecd2c2f940eb11a575140e0e2c267aa46757765#diff-e0a3628cb03c02182cfe6b4c1295da09R76). This array is used because a product normally has an array of categories which should be resolved simultaneously. This array is passed to the function to resolve the categories